lectures/sympy.md

```diff
@@ -0,0 +1,924 @@
+# Enable the best printer available
+init_printing()
+```
+
++++ {"user_expressions": []}
+
+## Symbolic algebra
+
+### Symbols
+
++++ {"user_expressions": []}
+
+Before we start manipulating the symbols, we initialize some symbols to work with
+
+```{code-cell} ipython3
+x, y, z = symbols('x y z')
+```
+
++++ {"user_expressions": []}
+
+Symbols are the basic units for symbolic computation in SymPy.
+
++++ {"user_expressions": []}
+
+### Expressions
+
+We can now use symbols `x`, `y`, and `z` to build expressions and equations.
+
+Here we build a simple expression first
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 10%
+---
+expr = (x + y) ** 2
+expr
+```
+
++++ {"user_expressions": []}
+
+We can expand this expression with the `expand` function
+
+```{code-cell} ipython3
+expand_expr = expand(expr)
+expand_expr
+```
+
++++ {"user_expressions": []}
+
+and factorize it back to the factored form with the `factor` function
+
+```{code-cell} ipython3
+factor(expand_expr)
+```
+
++++ {"user_expressions": []}
+
+We can solve this expression
+
+```{code-cell} ipython3
+solve(expr)
+```
+
++++ {"user_expressions": []}
+
+Note this is equivalent to solving the equation
+
+$$
+(x + y)^2 = 0 
+$$
+
+Another way to solve this equation is to use `solveset`.
+
+`solveset` allows users to define the domain of `x` in the evaluation and output a set as a solution
+
+```{code-cell} ipython3
+solveset(expr, x, domain=S.Naturals)
+```
+
++++ {"user_expressions": []}
+
+```{note}
+[Solvers](https://docs.sympy.org/latest/modules/solvers/index.html) is an important module with tools to solve different types of equations. 
+
+There are a variety of solvers available in SymPy depending on the nature of the problem.
+```
+
++++ {"user_expressions": []}
+
+### Equations
+
+SymPy provides several functions to manipulate equations.
+
+Let's develop an equation with the expression we defined before
+
+```{code-cell} ipython3
+eq = Eq(expr, 0)
+eq
+```
+
++++ {"user_expressions": []}
+
+Solving this equation with respect to x gives the same output as solving the expression directly
+
+```{code-cell} ipython3
+solve(eq, x)
+```
+
++++ {"user_expressions": []}
+
+SymPy can easily handle equations with multiple solutions
+
+```{code-cell} ipython3
+eq = Eq(expr, 1)
+solve(eq, x)
+```
+
++++ {"user_expressions": []}
+
+`solve` function can also combine multiple equations together and solve a system of equations
+
+```{code-cell} ipython3
+eq2 = Eq(x, y)
+eq2
+```
+
+```{code-cell} ipython3
+solve([eq, eq2], [x, y])
+```
+
++++ {"user_expressions": []}
+
+We can also solve for the value of $y$ by simply substituting $x$ with $y$
+
+```{code-cell} ipython3
+expr_sub = expr.subs(x, y)
+expr_sub
+```
+
+```{code-cell} ipython3
+solve(Eq(expr_sub, 1))
+```
+
++++ {"user_expressions": []}
+
+Below is another example equation with symbol `x` and functions `sin`, `cos`, and `tan` using the `Eq` function
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 17.5%
```

Comment:
There seems to be a lot of customisation to these sizes in `jb`

Reply:
I think @HengchengZhang made these changes to make the formula look better in the build. I was thinking about applying these configurations across the lecture.

Reply:
Hi @mmcky and @HumphreyYang, I think sympy doesn’t really have size config for outputs so I add these to make them look better.